driver/src/orderbook/streamed/block_timestamp.rs

```diff
@@ -21,3 +21,37 @@ impl BlockTimestamp for Web3 {
         r.boxed()
     }
 }
+
+/// A cache for the block timestamp which avoids having to query the node in the case where we
+/// receive multiple events from the same block in a row.
+#[derive(Debug)]
+pub struct BlockTimestampCache<T> {
+    block_timestamp: T,
+    hash: H256,
+    timestamp: u64,
+}
+
+impl<T> BlockTimestampCache<T> {
+    pub fn new(block_timestamp: T) -> Self {
+        Self {
+            block_timestamp,
+            hash: H256::zero(),
+            timestamp: 0,
+        }
+    }
+}
+
+// It would be nicer if we implemented `BlockTimestamp for BlockTimestampCache` but we cannot
+// because `async` is not yet supported in traits which in turn means we would have to return a
+// BoxFuture even on cache hit and it makes it harder to update the cache after waiting for the
+// `self.block_timestamp.block_timestamp` future.
+impl<T: BlockTimestamp> BlockTimestampCache<T> {
+    pub async fn block_timestamp(&mut self, block_hash: H256) -> Result<u64> {
+        if self.hash != block_hash {
+            let timestamp = self.block_timestamp.block_timestamp(block_hash).await?;
+            self.hash = block_hash;
+            self.timestamp = timestamp;
+        }
+        Ok(self.timestamp)
```

Comment:
Shouldn't you just be able to box an async block like this:

```suggestion
        async move {
            if self.hash != block_hash {
                let timestamp = self.block_timestamp.block_timestamp(block_hash).await?;
                self.hash = block_hash;
                self.timestamp = timestamp;
            }
            Ok(self.timestamp)
        }.boxed()
```

Reply:
Maybe the `move` is not necessary, but instinctively it is so that the future's lifetime is tied to `self`.

Also I noticed that the trait doesn't accept a mutable reference. Additionally, you might require to explicitely set lifetimes:
```
fn block_timestamp(&'a mut self, block_hash: H256) -> BoxFuture<'a, Result<u64>>;
```

Reply:
Thank you. It worked with your first suggestion and I changed the trait to `&mut self`.